src/virtio/net.c

```diff
@@ -0,0 +1,385 @@
+    assert(success);
+}
+
+static void handle_tx_msg(struct virtio_device *dev,
+                          struct virtq *virtq,
+                          uint16_t desc_head,
+                          bool *notify_tx_server,
+                          bool *respond_to_client)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (net_queue_full_active(&state->tx)) {
+        LOG_NET_WRN("TX used ring full\n");
+        goto fail;
+    }
+
+    net_buff_desc_t sddf_buffer;
+    int error = net_dequeue_free(&state->tx, &sddf_buffer);
+    if (error) {
+        LOG_NET_WRN("Avail ring is empty\n");
+        goto fail;
+    }
+
+    void *dest_buf = state->tx_data + sddf_buffer.io_or_offset;
+
+    uint32_t written = 0;
+    uint32_t dest_remaining = NET_BUFFER_SIZE;
+
+    /* Strip virtio header before copying to sDDF */
+    uint32_t skip_remaining = sizeof(struct virtio_net_hdr_mrg_rxbuf);
+
+    struct virtq_desc *desc = &virtq->desc[desc_head];
+
+    while (dest_remaining > 0) {
+        uint32_t skipping = 0;
+        /* Work out how much of this descriptor must be skipped */
+        skipping = MIN(skip_remaining, desc->len);
+        /* Truncate packets that are large than BUF_SIZE */
+        uint32_t writing = MIN(dest_remaining, desc->len - skipping);
+
+        memcpy(dest_buf + written, (void *)desc->addr + skipping, writing);
+
+        skip_remaining -= skipping;
+        written += writing;
+        dest_remaining -= writing;
+
+        if (desc->flags & VIRTQ_DESC_F_NEXT) {
+            desc = &virtq->desc[desc->next];
+        } else {
+            break;
+        }
+    }
+
+    sddf_buffer.len = written;
+    error = net_enqueue_active(&state->tx, sddf_buffer);
+    /* This cannot fail as we check above */
+    assert(!error);
+
+    virtq_enqueue_used(virtq, desc_head, written);
+    *respond_to_client = true;
+    *notify_tx_server = true;
+    return;
+
+fail:
+    virtq_enqueue_used(virtq, desc_head, 0);
+    *respond_to_client = true;
+}
+
+static int virtio_net_queue_notify(struct virtio_device *dev)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (dev->data.QueueSel != VIRTIO_NET_TX_VIRTQ) {
+        LOG_NET_ERR("Invalid queue\n");
+        return 0;
+    }
+
+    virtio_queue_handler_t *vq = &dev->vqs[VIRTIO_NET_TX_VIRTQ];
+    struct virtq *virtq = &vq->virtq;
+
+    uint16_t guest_idx = virtq->avail->idx;
+    uint16_t idx = vq->last_idx;
+
+    bool notify_tx_server = false;
+    bool respond_to_client = false;
+
+    for (; idx != guest_idx; idx++) {
+        uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+        handle_tx_msg(dev, virtq, desc_head, &notify_tx_server, &respond_to_client);
+    }
+
+    vq->last_idx = idx;
+
+    if (notify_tx_server && net_require_signal_active(&state->tx)) {
+        net_cancel_signal_active(&state->tx);
+        if (!have_signal) {
+            microkit_notify_delayed(state->tx_ch);
```

Comment:
This is subtle but the delayed API is used when you want to combine the next libmicrokit `Recv` call with a `Send` to save one system call.

In the mainline Microkit, when the fault handler finishes (which will indirectly invoke this function first) it will result in a `ReplyRecv`, which means this delay would not actually result in the signal happening. Right now libvmm does an explicit fault reply within the fault handler so there is no problem.

Because we will be switching to the mainline Microkit soon, I think that it would be best to change this to `microkit_notify` instead.

Reply:
This also shows that the `*_delayed` API is somewhat fragile and easy to make mistakes with. It hasn't been mainlined yet so would be worth thinking about how it could be better first. Anyways, that's not something to do with this PR.

Reply:
Yeah I had no idea what delayed was for. If we keep it it should be hidden away from the user as much as possible. I've removed it in https://github.com/au-ts/libvmm/pull/95/commits/37fc6dfd17a5b03ebb54404fd269d58be182706b